lib/phase/functional.js

```diff
@@ -168,6 +192,30 @@ function validateWorkflow(doc) {
 }
 
 /**
+ * Ensure the workflowGraph is a valid one
+ *  - Nodes contains all defined jobs (no extra or missing)
+ *  - No cycle in the graph
+ * @method validateWorkflowGraph
+ * @param  {Object}          doc Document that went through flattening
+ * @return {Array}               List of errors
+ */
+function validateWorkflowGraph(doc) {
+    // Validate that workflowGraph.nodes contains all jobs in `jobs`
+    const sortedJobList = JSON.stringify(clone(Object.keys(doc.jobs)).sort());
+    const nodesNames = clone(doc.workflowGraph.nodes).map(node => node.name);
+    const filteredNodes = nodesNames.filter(name => name !== '~commit' && name !== '~pr');
+    const sortedNodes = JSON.stringify(filteredNodes.sort());
+
+    if (sortedNodes !== sortedJobList) {
+        return ['WorkflowGraph: must contain all the jobs listed in "jobs" in nodes'];
+    }
+
+    // TODO: check if there is cycle in the graph
```

Comment:
`WorkflowParser.hasCycle(workflowGraph) // true/false`

Reply:
Petey, I remove the first check to see if nodes contain all jobs. The err return here is for users like they put things wrong in `yaml`. But that check is like checking the functionality of workflow-parser.